doc/TAM/SAI-Proposal-TAM-stream-telemetry.md

```diff
@@ -0,0 +1,906 @@
+
+// ...
+
+create_switch(&gSwitchId, (uint32_t)sai_attr_list.size(), sai_attr_list.data());
+
+```
+
+#### Creating HOSTIF object
+
+``` c++
+
+sai_attr_list[0].id = SAI_HOSTIF_ATTR_TYPE;
+sai_attr_list[0].value.s32 = SAI_HOSTIF_TYPE_GENETLINK;
+
+sai_attr_list[1].id = SAI_HOSTIF_ATTR_OPER_STATUS;
+sai_attr_list[1].value.boolean = true;
+
+// Set genetlink family
+sai_attr_list[2].id = SAI_HOSTIF_ATTR_NAME;
+strncpy(sai_attr_list[2].value.chardata, "stel", strlen("stel") + 1);
+
+// Set genetlink group
+sai_attr_list[3].id = SAI_HOSTIF_ATTR_GENETLINK_MCGRP_NAME;
+strncpy(sai_attr_list[3].value.chardata, "ipfix", strlen("ipfix") + 1);
+
+attr_count = 4;
+create_hostif(sai_hostif_obj, switch_id, attr_count, sai_attr_list);
+
+```
+
+#### Creating HOSTIF trap group
+
+``` c++
+
+create_hostif_trap_group(sai_trap_group_obj, switch_id, 0, NULL);
+
+```
+
+#### Creating HOSTIF user defined trap
+
+``` c++
+
+sai_attr_list[0].id = SAI_HOSTIF_USER_DEFINED_TRAP_ATTR_TYPE;
+sai_attr_list[0].value.s32 = SAI_HOSTIF_USER_DEFINED_TRAP_TYPE_TAM;
+
+sai_attr_list[1].id = SAI_HOSTIF_USER_DEFINED_TRAP_ATTR_TRAP_GROUP;
+sai_attr_list[1].value.oid = sai_trap_group_obj;
+
+attr_count = 2;
+sai_create_hostif_user_defined_trap_fn(&sai_hostif_udt_obj, switch_id, attr_count, sai_attr_list);
+
+```
+
+#### Creating Hostif table entry
+
+``` c++
+
+sai_attr_list[0].id = SAI_HOSTIF_TABLE_ENTRY_ATTR_TYPE;
+sai_attr_list[0].value.s32 = SAI_HOSTIF_TABLE_ENTRY_TYPE_TRAP_ID;
+
+sai_attr_list[1].id = SAI_HOSTIF_TABLE_ENTRY_ATTR_TRAP_ID;
+sai_attr_list[1].value.oid = sai_hostif_udt_obj;
+
+sai_attr_list[2].id = SAI_HOSTIF_TABLE_ENTRY_ATTR_CHANNEL_TYPE;
+sai_attr_list[2].value.s32 = SAI_HOSTIF_TABLE_ENTRY_CHANNEL_TYPE_GENETLINK;
+
+sai_attr_list[3].id = SAI_HOSTIF_TABLE_ENTRY_ATTR_HOST_IF;
+sai_attr_list[3].value.oid = sai_hostif_obj;
+
+attr_count = 4;
+sai_create_hostif_table_entry_fn(&sai_hostif_table_entry_obj, switch_id, attr_count, sai_attr_list);
+
+```
+
+#### Creating TAM transport object
+
+``` c++
+
+sai_attr_list[0].id = SAI_TAM_TRANSPORT_ATTR_TRANSPORT_TYPE;
+sai_attr_list[0].value.s32 = SAI_TAM_TRANSPORT_TYPE_NONE;
+attr_count = 1;
+sai_create_tam_transport_fn(&sai_tam_transport_obj, switch_id, attr_count, sai_attr_list);
+
+```
+
+#### Creating TAM collector object
+
+``` c++
+
+sai_attr_list[0].id = SAI_TAM_COLLECTOR_ATTR_TRANSPORT;
+sai_attr_list[0].value.oid = sai_tam_transport_obj;
+
+sai_attr_list[1].id = SAI_TAM_COLLECTOR_ATTR_LOCALHOST;
+sai_attr_list[1].value.booldata = true;
+
+sai_attr_list[2].id = SAI_TAM_COLLECTOR_ATTR_HOSTIF_TRAP;
+sai_attr_list[2].value.oid = sai_hostif_udt_obj;
+
+sai_attr_list[3].id = SAI_TAM_COLLECTOR_ATTR_DSCP_VALUE;
+sai_attr_list[3].value.u8 = 0;
+
+attr_count = 4;
+sai_create_tam_collector_fn(&sai_tam_collector_obj, switch_id, attr_count, sai_attr_list);
+
+```
+
+#### Creating TAM report object
+
+``` c++
+
+sai_attr_list[0].id = SAI_TAM_REPORT_ATTR_TYPE;
+sai_attr_list[0].value.s32 = SAI_TAM_REPORT_TYPE_IPFIX;

```

Comment:
The report type should be GENETLINK, and the encoding within the GENETLINK is IPFIX. The latter part is already specified using SAI\_HOSTIF\_ATTR\_GENETLINK\_MCGRP\_NAME.

Reply:
IMHO, this attribute is for specifying the format of message, Other values of this attribute is more like the format instead of channel.
![image](https://github.com/user-attachments/assets/ff753b64-35e0-4934-ad03-714f63788d4a)

Reply:
And I didn't see an attribute for encoding format if not to specify at here.